File: clients/java/Fylo.java
// Fylo client — drives the `fylo` binary's persistent NDJSON loop.
//
// No dependencies (java.lang.Process only). Requires the `fylo` binary on PATH
// (brew/scoop) or an explicit path. One long-lived subprocess keeps the engine
// warm across calls.
//
//   try (Fylo db = new Fylo("/path/to/db")) {
//       db.createCollection("users");
//       String put = db.putData("users", Map.of("name", "Ada", "role", "admin"));
//       String admins = db.findDocs("users",
//           Map.of("$ops", List.of(Map.of("role", Map.of("$eq", "admin")))));
//   }
//
// Each operation method builds the request, checks it succeeded, and returns the
// raw JSON response line (parse `result` with Jackson/Gson). Method names follow
// Java's camelCase convention; object arguments are native Maps/Lists, encoded
// to JSON by the built-in `toJson`. request(json) is the raw escape hatch.

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public final class Fylo implements AutoCloseable {
    private final Process proc;
    private final BufferedWriter in;
    private final BufferedReader out;

    public Fylo(String root) throws IOException {
        this(root, "fylo", false);
    }

    public Fylo(String root, String binary, boolean worm) throws IOException {
        List<String> args = new ArrayList<>(List.of(binary, "exec", "--loop", "--root", root));
        if (worm) args.add("--worm");
        this.proc = new ProcessBuilder(args)
                .redirectError(ProcessBuilder.Redirect.INHERIT)
                .start();
        this.in = new BufferedWriter(
                new OutputStreamWriter(proc.getOutputStream(), StandardCharsets.UTF_8));
        this.out = new BufferedReader(
                new InputStreamReader(proc.getInputStream(), StandardCharsets.UTF_8));
    }

    /** Send one raw machine-protocol op (JSON string); returns the response line. */
    public synchronized String request(String opJson) throws IOException {
        if (!proc.isAlive()) throw new IOException("fylo process has exited");
        in.write(opJson.stripTrailing());
        in.write('\n');
        in.flush();
        String line = out.readLine();
        if (line == null) throw new IOException("fylo closed the stream");
        return line;
    }

    // Build an op from native fields, send it, and error on a failure response.
    // ponytail: checks for the always-present "ok":true field by substring.
    private String op(String name, Object... kv) throws IOException {
        StringBuilder sb = new StringBuilder("{\"op\":").append(toJson(name));
        for (int i = 0; i + 1 < kv.length; i += 2) {
            sb.append(',').append(toJson(kv[i].toString())).append(':').append(toJson(kv[i + 1]));
        }
        String resp = request(sb.append('}').toString());
        if (!resp.contains("\"ok\":true")) throw new IOException(resp.strip());
        return resp;
    }

    // Quote a string as a JSON string literal, escaping control characters so an
    // embedded newline/tab can't break the newline-delimited protocol.
    static String quote(String s) {
        StringBuilder sb = new StringBuilder(s.length() + 2).append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"': sb.append("\\\""); break;
                case '\\': sb.append("\\\\"); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                case '\b': sb.append("\\b"); break;
                case '\f': sb.append("\\f"); break;
                default:
                    if (c < 0x20) sb.append(String.format("\\u%04x", (int) c));
                    else sb.append(c);
            }
        }
        return sb.append('"').toString();
    }

    // Minimal JSON encoder for String / Number / Boolean / Map / Iterable / null.
    static String toJson(Object v) {
        if (v == null) return "null";
        if (v instanceof String) {
            return quote((String) v);
        }
        if (v instanceof Boolean || v instanceof Number) return v.toString();
        if (v instanceof Map) {
            StringBuilder sb = new StringBuilder("{");
            boolean first = true;
            for (Map.Entry<?, ?> e : ((Map<?, ?>) v).entrySet()) {
                if (!first) sb.append(',');
                first = false;
                sb.append(toJson(e.getKey().toString())).append(':').append(toJson(e.getValue()));
            }
            return sb.append('}').toString();
        }
        if (v instanceof Iterable) {
            StringBuilder sb = new StringBuilder("[");
            boolean first = true;
            for (Object x : (Iterable<?>) v) {
                if (!first) sb.append(',');
                first = false;
                sb.append(toJson(x));
            }
            return sb.append(']').toString();
        }
        throw new IllegalArgumentException("unsupported JSON value: " + v.getClass());
    }

    // --- Collections ---
    public String createCollection(String collection) throws IOException {
        return createCollection(collection, "document");
    }
    public String createCollection(String collection, String kind) throws IOException {
        return op("createCollection", "collection", collection, "kind", kind);
    }
    public String dropCollection(String collection) throws IOException {
        return op("dropCollection", "collection", collection);
    }
    public String inspectCollection(String collection) throws IOException {
        return op("inspectCollection", "collection", collection);
    }
    public String rebuildCollection(String collection) throws IOException {
        return op("rebuildCollection", "collection", collection);
    }

    // --- Documents (object args are native Maps/Lists) ---
    public String putData(String collection, Map<String, Object> data) throws IOException {
        return op("putData", "collection", collection, "data", data);
    }
    public String getDoc(String collection, String id) throws IOException {
        return op("getDoc", "collection", collection, "id", id);
    }
    public String getMeta(String collection, String id) throws IOException {
        return op("getMeta", "collection", collection, "id", id);
    }
    public String setMeta(String collection, String id, Map<String, Object> meta) throws IOException {
        return op("setMeta", "collection", collection, "id", id, "meta", meta);
    }
    public String getLatest(String collection, String id) throws IOException {
        return op("getLatest", "collection", collection, "id", id);
    }
    public String patchDoc(String collection, String id, Map<String, Object> newDoc)
            throws IOException {
        return op("patchDoc", "collection", collection, "id", id, "newDoc", newDoc);
    }
    public String delDoc(String collection, String id) throws IOException {
        return op("delDoc", "collection", collection, "id", id);
    }
    public String restoreDoc(String collection, String id) throws IOException {
        return op("restoreDoc", "collection", collection, "id", id);
    }

    // --- Query ---
    public String findDocs(String collection, Map<String, Object> query) throws IOException {
        return op("findDocs", "collection", collection, "query", query);
    }
    public String executeSQL(String sql) throws IOException {
        return op("executeSQL", "sql", sql);
    }

    // Run raw SQL, built with concatenation/String.format. Values are inlined
    // verbatim — escape/validate untrusted input yourself.
    public String sql(String query) throws IOException {
        return executeSQL(query);
    }

    /** Collection-scoped facade: db.collection("users").put(data). */
    public Collection collection(String name) {
        return new Collection(name);
    }

    /** A collection-scoped view; methods drop the leading collection argument. */
    public final class Collection {
        private final String name;

        private Collection(String name) {
            this.name = name;
        }

        public String create() throws IOException {
            return createCollection(name);
        }
        public String create(String kind) throws IOException {
            return createCollection(name, kind);
        }
        public String drop() throws IOException {
            return dropCollection(name);
        }
        public String inspect() throws IOException {
            return inspectCollection(name);
        }
        public String rebuild() throws IOException {
            return rebuildCollection(name);
        }
        public String put(Map<String, Object> data) throws IOException {
            return putData(name, data);
        }
        public String get(String id) throws IOException {
            return getDoc(name, id);
        }
        public String getMeta(String id) throws IOException { return Fylo.this.getMeta(name, id); }
        public String setMeta(String id, Map<String, Object> meta) throws IOException {
            return Fylo.this.setMeta(name, id, meta);
        }
        public String latest(String id) throws IOException {
            return getLatest(name, id);
        }
        public String patch(String id, Map<String, Object> newDoc) throws IOException {
            return patchDoc(name, id, newDoc);
        }
        public String delete(String id) throws IOException {
            return delDoc(name, id);
        }
        public String restore(String id) throws IOException {
            return restoreDoc(name, id);
        }
        public String find(Map<String, Object> query) throws IOException {
            return findDocs(name, query);
        }
    }

    @Override
    public void close() throws IOException {
        if (proc.isAlive()) {
            in.close(); // EOF ends the loop
            try {
                proc.waitFor();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}

File: clients/rust/fylo.rs
//! Fylo client — drives the `fylo` binary's persistent NDJSON loop.
//!
//! No crates (std only), so it works as a single-file module or dropped into a
//! crate. Requires the `fylo` binary on PATH (brew/scoop) or an explicit path.
//! One long-lived child process keeps the engine warm across calls.
//!
//! ```no_run
//! use fylo::{Fylo, Json};
//! let mut db = Fylo::open("/path/to/db", "fylo", false).unwrap();
//! db.create_collection("users", "document").unwrap();
//! db.put_data("users", Json::obj(vec![("name", "Ada".into()), ("role", "admin".into())])).unwrap();
//! // responses are raw JSON lines: {"ok":true,"result":"<id>",...}
//! let admins = db.find_docs("users",
//!     Json::obj(vec![("$ops", Json::arr(vec![
//!         Json::obj(vec![("role", Json::obj(vec![("$eq", "admin".into())]))])]))])).unwrap();
//! ```
//!
//! Operation methods build the request for you and error on `"ok":false`; they
//! return the raw JSON response line (bring serde if you want typed structs).
//! Object arguments are built with the dependency-free `Json` value type (which
//! has `From` impls for &str/String/i64/f64/bool). Method names follow Rust's
//! snake_case; `request` is the raw escape hatch for ops without a method.

use std::io::{BufRead, BufReader, Write};
use std::process::{Child, ChildStdin, Command, Stdio};

pub struct Fylo {
    child: Child,
    stdin: Option<ChildStdin>,
    stdout: BufReader<std::process::ChildStdout>,
}

impl Fylo {
    /// Start a warm fylo process rooted at `root`. `binary` is usually "fylo".
    pub fn open(root: &str, binary: &str, worm: bool) -> std::io::Result<Fylo> {
        let mut cmd = Command::new(binary);
        cmd.args(["exec", "--loop", "--root", root]);
        if worm {
            cmd.arg("--worm");
        }
        let mut child = cmd.stdin(Stdio::piped()).stdout(Stdio::piped()).spawn()?;
        let stdin = child.stdin.take().expect("stdin piped");
        let stdout = BufReader::new(child.stdout.take().expect("stdout piped"));
        Ok(Fylo { child, stdin: Some(stdin), stdout })
    }

    /// Send one machine-protocol operation (a JSON object string) and return the
    /// response line (also JSON). ponytail: one call in flight; not thread-safe.
    pub fn request(&mut self, op_json: &str) -> std::io::Result<String> {
        let stdin = self
            .stdin
            .as_mut()
            .ok_or_else(|| std::io::Error::new(std::io::ErrorKind::BrokenPipe, "fylo closed"))?;
        stdin.write_all(op_json.trim_end().as_bytes())?;
        stdin.write_all(b"\n")?;
        stdin.flush()?;
        let mut line = String::new();
        let n = self.stdout.read_line(&mut line)?;
        if n == 0 {
            return Err(std::io::Error::new(
                std::io::ErrorKind::UnexpectedEof,
                "fylo closed the stream",
            ));
        }
        Ok(line)
    }

    // Send a fully-formed op JSON and error on a failure response.
    // ponytail: checks for the always-present "ok":true field by substring.
    fn checked(&mut self, json: String) -> std::io::Result<String> {
        let resp = self.request(&json)?;
        if !resp.contains("\"ok\":true") {
            return Err(std::io::Error::new(std::io::ErrorKind::Other, resp.trim().to_string()));
        }
        Ok(resp)
    }

    // --- Collections ---
    pub fn create_collection(&mut self, collection: &str, kind: &str) -> std::io::Result<String> {
        let kind = if kind.is_empty() { "document" } else { kind };
        self.checked(format!(
            r#"{{"op":"createCollection","collection":"{}","kind":"{}"}}"#,
            esc(collection),
            esc(kind)
        ))
    }
    pub fn drop_collection(&mut self, collection: &str) -> std::io::Result<String> {
        self.checked(format!(r#"{{"op":"dropCollection","collection":"{}"}}"#, esc(collection)))
    }
    pub fn inspect_collection(&mut self, collection: &str) -> std::io::Result<String> {
        self.checked(format!(r#"{{"op":"inspectCollection","collection":"{}"}}"#, esc(collection)))
    }
    pub fn rebuild_collection(&mut self, collection: &str) -> std::io::Result<String> {
        self.checked(format!(r#"{{"op":"rebuildCollection","collection":"{}"}}"#, esc(collection)))
    }

    // --- Documents (object args are built with Json) ---
    pub fn put_data(&mut self, collection: &str, data: Json) -> std::io::Result<String> {
        self.checked(format!(
            r#"{{"op":"putData","collection":"{}","data":{}}}"#,
            esc(collection),
            data.encode()
        ))
    }
    pub fn get_doc(&mut self, collection: &str, id: &str) -> std::io::Result<String> {
        self.checked(format!(
            r#"{{"op":"getDoc","collection":"{}","id":"{}"}}"#,
            esc(collection),
            esc(id)
        ))
    }
    pub fn get_meta(&mut self, collection: &str, id: &str) -> std::io::Result<String> {
        self.checked(format!(
            r#"{{"op":"getMeta","collection":"{}","id":"{}"}}"#,
            esc(collection),
            esc(id)
        ))
    }
    pub fn set_meta(&mut self, collection: &str, id: &str, meta: Json) -> std::io::Result<String> {
        self.checked(format!(
            r#"{{"op":"setMeta","collection":"{}","id":"{}","meta":{}}}"#,
            esc(collection),
            esc(id),
            meta.encode()
        ))
    }
    pub fn get_latest(&mut self, collection: &str, id: &str) -> std::io::Result<String> {
        self.checked(format!(
            r#"{{"op":"getLatest","collection":"{}","id":"{}"}}"#,
            esc(collection),
            esc(id)
        ))
    }
    pub fn patch_doc(&mut self, collection: &str, id: &str, new_doc: Json) -> std::io::Result<String> {
        self.checked(format!(
            r#"{{"op":"patchDoc","collection":"{}","id":"{}","newDoc":{}}}"#,
            esc(collection),
            esc(id),
            new_doc.encode()
        ))
    }
    pub fn del_doc(&mut self, collection: &str, id: &str) -> std::io::Result<String> {
        self.checked(format!(
            r#"{{"op":"delDoc","collection":"{}","id":"{}"}}"#,
            esc(collection),
            esc(id)
        ))
    }
    pub fn restore_doc(&mut self, collection: &str, id: &str) -> std::io::Result<String> {
        self.checked(format!(
            r#"{{"op":"restoreDoc","collection":"{}","id":"{}"}}"#,
            esc(collection),
            esc(id)
        ))
    }

    // --- Query ---
    pub fn find_docs(&mut self, collection: &str, query: Json) -> std::io::Result<String> {
        self.checked(format!(
            r#"{{"op":"findDocs","collection":"{}","query":{}}}"#,
            esc(collection),
            query.encode()
        ))
    }
    pub fn execute_sql(&mut self, sql: &str) -> std::io::Result<String> {
        self.checked(format!(r#"{{"op":"executeSQL","sql":"{}"}}"#, esc(sql)))
    }

    /// Run raw SQL, built with `format!`. Values are inlined verbatim —
    /// escape/validate untrusted input yourself.
    pub fn sql(&mut self, query: &str) -> std::io::Result<String> {
        self.execute_sql(query)
    }

    /// Collection-scoped facade with short method names, so
    /// `db.collection("users").put(data)` reads like the browser client.
    pub fn collection<'a>(&'a mut self, name: &str) -> Collection<'a> {
        Collection { db: self, name: name.to_string() }
    }
}

/// A collection-scoped view; methods drop the leading collection argument.
pub struct Collection<'a> {
    db: &'a mut Fylo,
    name: String,
}

impl<'a> Collection<'a> {
    pub fn create(&mut self, kind: &str) -> std::io::Result<String> {
        self.db.create_collection(&self.name, kind)
    }
    pub fn drop(&mut self) -> std::io::Result<String> {
        self.db.drop_collection(&self.name)
    }
    pub fn inspect(&mut self) -> std::io::Result<String> {
        self.db.inspect_collection(&self.name)
    }
    pub fn rebuild(&mut self) -> std::io::Result<String> {
        self.db.rebuild_collection(&self.name)
    }
    pub fn put(&mut self, data: Json) -> std::io::Result<String> {
        self.db.put_data(&self.name, data)
    }
    pub fn get(&mut self, id: &str) -> std::io::Result<String> {
        self.db.get_doc(&self.name, id)
    }
    pub fn get_meta(&mut self, id: &str) -> std::io::Result<String> {
        self.db.get_meta(&self.name, id)
    }
    pub fn set_meta(&mut self, id: &str, meta: Json) -> std::io::Result<String> {
        self.db.set_meta(&self.name, id, meta)
    }
    pub fn latest(&mut self, id: &str) -> std::io::Result<String> {
        self.db.get_latest(&self.name, id)
    }
    pub fn patch(&mut self, id: &str, new_doc: Json) -> std::io::Result<String> {
        self.db.patch_doc(&self.name, id, new_doc)
    }
    pub fn delete(&mut self, id: &str) -> std::io::Result<String> {
        self.db.del_doc(&self.name, id)
    }
    pub fn restore(&mut self, id: &str) -> std::io::Result<String> {
        self.db.restore_doc(&self.name, id)
    }
    pub fn find(&mut self, query: Json) -> std::io::Result<String> {
        self.db.find_docs(&self.name, query)
    }
}

// Minimal JSON string escaping for interpolated scalar values.
fn esc(s: &str) -> String {
    let mut out = String::with_capacity(s.len() + 2);
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            '\u{08}' => out.push_str("\\b"),
            '\u{0c}' => out.push_str("\\f"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out
}

/// A tiny dependency-free JSON value for building object arguments natively.
/// Scalars convert via `.into()` (e.g. `"admin".into()`, `18.into()`).
pub enum Json {
    Null,
    Bool(bool),
    Num(f64),
    Str(String),
    Arr(Vec<Json>),
    Obj(Vec<(String, Json)>),
}

impl Json {
    pub fn arr(items: Vec<Json>) -> Json {
        Json::Arr(items)
    }
    pub fn obj(pairs: Vec<(&str, Json)>) -> Json {
        Json::Obj(pairs.into_iter().map(|(k, v)| (k.to_string(), v)).collect())
    }

    fn encode(&self) -> String {
        match self {
            Json::Null => "null".to_string(),
            Json::Bool(b) => b.to_string(),
            Json::Num(n) if n.fract() == 0.0 => format!("{}", *n as i64),
            Json::Num(n) => n.to_string(),
            Json::Str(s) => format!("\"{}\"", esc(s)),
            Json::Arr(a) => {
                let items: Vec<String> = a.iter().map(|x| x.encode()).collect();
                format!("[{}]", items.join(","))
            }
            Json::Obj(o) => {
                let pairs: Vec<String> =
                    o.iter().map(|(k, v)| format!("\"{}\":{}", esc(k), v.encode())).collect();
                format!("{{{}}}", pairs.join(","))
            }
        }
    }
}

impl From<&str> for Json {
    fn from(s: &str) -> Json {
        Json::Str(s.to_string())
    }
}
impl From<String> for Json {
    fn from(s: String) -> Json {
        Json::Str(s)
    }
}
impl From<i64> for Json {
    fn from(n: i64) -> Json {
        Json::Num(n as f64)
    }
}
impl From<f64> for Json {
    fn from(n: f64) -> Json {
        Json::Num(n)
    }
}
impl From<bool> for Json {
    fn from(b: bool) -> Json {
        Json::Bool(b)
    }
}

impl Drop for Fylo {
    fn drop(&mut self) {
        // Close stdin FIRST so the loop hits EOF and exits, then reap the child.
        self.stdin.take();
        let _ = self.child.wait();
    }
}
